sklearn/utils/estimator_checks.py

```diff
@@ -429,11 +430,17 @@ class NotAnArray:
     """
 
     def __init__(self, data):
-        self.data = data
+        self.data = np.asarray(data)
 
     def __array__(self, dtype=None):
         return self.data
 
+    def __array_function__(self, func, types, args, kwargs):
```

Comment:
Returning `True` and raising `TypeError` needs to be tested?

Reply:
you mean explicitly tested? Sure, i could do that. Thought that was a bit overkill for a test helper. They are obviously used in the tests, but on CI there's no ``__array_function__`` protocol.

Reply:
I would be more onboard if `NotAnArray` was private, which goes back to "public vs private utils" https://github.com/scikit-learn/scikit-learn/issues/6616#issuecomment-516479413

On the other hand, we are depending on this raising when something is wrong. When everything is working, our tests do not run `__array_function__`.

Reply:
That's a good point. I'll add a test (that won't be run)